utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt

```diff
@@ -0,0 +1,158 @@
+        } else {
+          listParents.pop()
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
+          throw EmptyStackException()
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
+    /**
+     * Opens a new list item.
+     *
+     * @param text
+     */
+    open fun openItem(text: Editable) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val len = text.length
+      text.setSpan(this, len, len, Spanned.SPAN_MARK_MARK)
+    }
+
+    /**
+     * Closes a list item.
+     *
+     * @param text
+     * @param indentation
+     */
+    fun closeItem(text: Editable, indentation: Int) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val replaces = getReplaces(text, indentation)
+      val len = text.length
+      val listTag = getLast(text)
+      val where = text.getSpanStart(listTag)
+      text.removeSpan(listTag)
+      if (where != len) {
+        for (replace in replaces) {
+          text.setSpan(replace, where, len, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+        }
+      }
+    }
+
+    protected abstract fun getReplaces(text: Editable?, indentation: Int): Array<Any>
+    /** Note: This knows that the last returned object from getSpans() will be the most recently added. */
+    private fun getLast(text: Spanned): ListItemTag? {
+      val listTags = text.getSpans(0, text.length, ListItemTag::class.java)
+      return if (listTags.size == 0) {
+        null
+      } else listTags[listTags.size - 1]
+    }
+  }
+
+  /** Class representing the unordered list (`<ul>`) HTML tag. */
+  private inner class UnorderedListTag : ListItemTag() {
+    override fun getReplaces(
+      text: Editable?,
+      indentation: Int
+    ): Array<Any> {
+      var bulletMargin: Int = indent
+      if (indentation > 1) {
+        bulletMargin = indent - bulletSpan.getLeadingMargin(true)
+        if (indentation > 2) {
+          bulletMargin -= (indentation - 2) * listItemIndent
+        }
+      }
+      return arrayOf(
+        LeadingMarginSpan.Standard(listItemIndent * (indentation - 1)),
+        BulletSpanWithRadius(context, bulletMargin)
+      )
+    }
+  }
+
+  /** Class representing the ordered list (`<ol>`) HTML tag. */
+  private inner class OrderedListTag
+  /** Creates a new `<ol>` with start index of 1. */
+  @JvmOverloads constructor(private var nextIdx: Int = 1) : ListItemTag() {
+    override fun openItem(text: Editable) {
+      super.openItem(text)
+      text.append(Integer.toString(nextIdx++)).append(". ")
```

Comment:
Suggest making this part of a separate abstract class that the open function calls rather than overriding it directly (e.g. ``getNextOpenItemPrefix``).